### ISSUE TITLE
s3 cache: allow to configure the endpoint

### ISSUE THREAD

**Issue by landryb** [mapproxy/mapproxy#348]
## Context
I'm playing with s3 cache, and for dev purposes i use minio instead of having a real fullblown aws account. Right now as mapproxy directly uses boto3 sessions, it seems there's no way to configure the endpoint_url, which can only be passed to the 'low-level' client creation (cf https://docs.minio.io/docs/how-to-use-aws-sdk-for-python-with-minio-server and http://boto3.readthedocs.io/en/latest/reference/core/session.html#boto3.session.Session.client) 

It would be nice to somehow be able to set the endpoint in the mapproxy config, as the boto3 and awscli devs dont seem to want to make it easy via env vars or default configuration file (per https://github.com/boto/boto3/issues/1375)

I'll see if i can come up with something..

**Comment by landryb** [comment]
I figured out i can make mapproxy connect to my minio by adding required args (key/secret/endpoint url) to the client constructure in https://github.com/mapproxy/mapproxy/blob/master/mapproxy/cache/s3.py#L85, now what would be the best way to make it configurable without breaking existing working setups ?

**Comment by stvno** [comment]
I have the same question, since I like to use mapproxy with a like-S3 service (Digital Ocean's Spaces), which has the same API as S3, but obviously different endpoints.

@landryb did you extend the line you mentioned something like this?
````
s3_session().client("s3", None, False, None, 'endpoint_url')
```
If so I can imagine that a check in Mapproxy Config for an endpoint setting might do the trick. (similar as my proposal for Tegola's support for non AWS S3 endpoints https://github.com/go-spatial/tegola/pull/414 )

@olt if you are happy with an extra check on the endpoint url I'm happy to write a PR

**Comment by olt** [comment]
There is a PR #364 for that, but I haven't found time to review it.
@stvno does this PR work for you?

**Comment by landryb** [comment]
Nice ! Will try to reserve some time to give this PR a shot against a minio server on localhost..

**Comment by landryb** [comment]
@stvno iirc i also added the credentials to the client() call, as i didnt know where to put them outside mapproxy (i dunno much about all those s3 things, but i suppose the boto lib uses .aws/ etc..)

**Comment by stvno** [comment]
@olt yes it makes it possible to upload files to a custom S3 server. However the default policy of Digital Ocean is to set the files private. I've created a new PR on top of @mikkomet's work to allow for ACLs #366

**Comment by olt** [comment]
Should be fixed with #366
